consumer_test.go

```diff
@@ -56,195 +53,491 @@ func TestConsumerOffsetManual(t *testing.T) {
-	// This way it is set correctly for the next FetchRequest.
-	if consumer.(*partitionConsumer).offset != 11 {
-		t.Error("Latest offset not fetched correctly:", consumer.(*partitionConsumer).offset)
+	// Then
+	if pc2 != nil || err != ConfigurationError("That topic/partition is already being consumed") {
+		t.Fatal("A partition cannot be consumed twice at the same time")
 	}
+
+	safeClose(t, pc1)
+	safeClose(t, c)
 }
 
-func TestConsumerShutsDownOutOfRange(t *testing.T) {
-	seedBroker := newMockBroker(t, 1)
-	leader := newMockBroker(t, 2)
+// If consumer fails to refresh metadata it keeps retrying every with frequency
+// given in `Config.Consumer.Retry.Backoff`.
+func TestConsumerLeaderRefreshError(t *testing.T) {
+	// Given
+	broker0 := newMockBroker(t, 100)
+	defer broker0.Close()
+
+	// Stage 1: my_topic/0 served by broker0
+	Logger.Printf("    STAGE 1")
+
+	broker0.SetHandlerByMap(map[string]MockResponse{
+		"MetadataRequest": newMockMetadataResponse(t).
+			SetBroker(broker0.Addr(), broker0.BrokerID()).
+			SetLeader("my_topic", 0, broker0.BrokerID()),
+		"OffsetRequest": newMockOffsetResponse(t).
+			SetOffset("my_topic", 0, OffsetOldest, 123).
+			SetOffset("my_topic", 0, OffsetNewest, 1000),
+		"FetchRequest": newMockFetchResponse(t, 1).
+			SetMessage("my_topic", 0, 123, testMsg),
+	})
+
+	config := NewConfig()
+	config.Net.ReadTimeout = 100 * time.Millisecond
+	config.Consumer.Retry.Backoff = 500 * time.Millisecond
+	config.Metadata.Retry.Max = 0
+	c, err := NewConsumer([]string{broker0.Addr()}, config)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	pc, err := c.ConsumePartition("my_topic", 0, OffsetOldest)
+	if err != nil {
+		t.Errorf("Failed to create a partition consumer, err=%v", err)
+	}
+
+	assertMessageOffset(t, <-pc.Messages(), 123)
+
+	// Stage 2: broker0 says that it is no longer the leader for my_topic/0,
+	// but the requests to retrieve metadata fail with network timeout.
+	Logger.Printf("    STAGE 2")
+
+	fetchResponse2 := &FetchResponse{}
+	fetchResponse2.AddError("my_topic", 0, ErrNotLeaderForPartition)
+
+	broker0.SetHandlerByMap(map[string]MockResponse{
+		"FetchRequest": newMockWrapper(fetchResponse2),
+	})
 
-	metadataResponse := new(MetadataResponse)
-	metadataResponse.AddBroker(leader.Addr(), leader.BrokerID())
-	metadataResponse.AddTopicPartition("my_topic", 0, leader.BrokerID(), nil, nil, ErrNoError)
-	seedBroker.Returns(metadataResponse)
+	// Stage 3: finally the metadata returned by broker0 tells that broker1 is
+	// a new leader for my_topic/0. Consumption resumes.
 
-	offsetResponseNewest := new(OffsetResponse)
-	offsetResponseNewest.AddTopicPartition("my_topic", 0, 1234)
-	leader.Returns(offsetResponseNewest)
+	// Unfortunately consumer does not propagate `ErrNotLeaderForPartition`
+	// error to PartitionConsumer.Errors() channel. So there is no other way to
+	// synchronize here by sleep.
+	time.Sleep(300 * time.Millisecond)
```

Comment:
the consumer *does* propagate the network timeouts to the errors channel, so you can synchronize on that right?